smsd/man/smsd.8
.TH "smsd" "8" "September 27, 2010" "" "Gnokii"
.SH "NAME"
smsd \- daemon for handling incoming and outgoing SMSes using libgnokii
.SH "SYNOPSIS"
.B smsd \fI[OPTIONS]\fR

.SH "DESCRIPTION"
.PP 
The
.B SMSD
(SMS daemon) program is intended for receiving and sending SMSes.

.PP
The program is designed to use modules (plugins) to work with different backends.
Currently are supported PostgreSQL, MySQL and a special module 'file' which is
designed to work without an SQL database (see below for special instructions
in the file module section). When \fBSMSD\fR uses an SQL server as a
repository for SMSes you must create a database (default name is 'sms') with
tables defined in sms.tables.{module}.sql files.

Table 'inbox' is for received SMSes. \fBSMSD\fR periodically checks a given
phone memory to move messages to this table.
Only received SMSes and status reports are deleted from the phone, other kinds
of messages are not removed.

.PP
Table 'inbox' has the following fields:

.RS
.TP
.B number
phone number from which is sms received
.TP
.B smsdate
date set by SMS center
.TP
.B insertdate
date SMS was inserted into table
.TP
.B text
SMS text
.TP
.B phone
number of phone to use. If you use smsd without \-t option you can ignore
this column. See \-t option below.
.TP
.B processed
smsd sets this always to FALSE. You can use it for another program which will parse SMSes.
.RE

.PP
Table 'outbox' is for outgoing SMSes. \fBSMSD\fR periodically looks into this table
and searches for lines with attribute 'processed' set to FALSE. If it finds
such SMS, it will send SMS to 'number'.

.PP
Table 'outbox' has the following fields:

.RS
.TP
.B number
recipient's phone number

.TP
.B processed_date
date of processing SMS.

.TP
.B insertdate
date of inserting SMS in table

.TP
.B text
SMS text

.TP
.B phone
number of phone to use. If you use smsd without \-t option you
must not fill this column. See \-t option below.

.TP
.B processed
new SMS must have this attribute set to FALSE. \fBSMSD\fR will send this SMS and will update 'processed' to TRUE.

.TP
.B error
This attribute contains return value of gn_sms_send() function. For explanation of numeric values see <gnokii/error.h> in gnokii source.

.TP
.B dreport
request for delivery report per message. Possible values are 0 for no delivery report, 1 for delivery report.

.TP
.B not_before
sms will not be sent before this time (default 00:00:00).

.TP
.B not_after
sms will not be sent after this time (default 23:59:59).
.RE
.PP
\fBSMSD\fR is multi threaded program. If you list processes you will see more than
one instance of smsd.

.SS
THE FILE MODULE
.PP
This module doesn't need an SQL server to work. For incoming messages it
works in a way that it calls a program giving phone number, date as parameter
and the message text on stdin. In the smsd directory there is an example file
action. It shows sms data to stdout. If you want something useful you must
write your own action.
.BR
For outgoing messages you can set spool directory where smsd looks
regularly.

.PP
Incoming sms are passed in format
.RS
.PD 0
.TP
.BR action\ number\ date\ <\ sms
.PD
.RE

.PP
Outgoing sms must be in format:
.RS
.PD 0
.TP
.B number
.TP
.B text
.PD
.RE

.PP
The name of file is unimportant but smsd ignores all files with prefix
"ERR.". If smsd successfully sends sms, it removes file from spool directory.
If smsd cannot send sms prepends file with "ERR." prefix and ignores it.
If you use \fI-m file\fR, \fI-u\fR (\fI--user\fR) and \fI-c\fR (\fI--host\fR) will have other meaning.
\fI-u\fR is for action (action must be executable)
\fI-c\fR is for spool directory where smsd looks for smses. Smsd must have read
and write permissions to this directory.

.PP
\fI-p\fR, \fI-d\fR parameters have no meaning when \fI-m file\fR is used.

.PP
If you don't specify \fI-u\fR, default action is to print sms to stdout.
If you don't specify \fI-c\fR, sms sending will be disabled.

.SS OPTIONS
.PP
Available options:
.TP
.BR -u,\ --user\ db_username
Specify user name.

.TP
.BR -p,\ --password\ db_password
Specify user password.

.TP
.BR -d,\ --db\ db_name
Specify database name for 'pq' and 'mysql' modules or full path to database file for 'sqlite' module (defaults to 'sms').

.TP
.BR -c,\ --host\ db_hostname
Specify database server host name.

.TP
.BR -s,\ --schema\ db_hostname
Specify database schema used by the PostgreSQL module (defaults to 'public').

.TP
.BR -e,\ --encoding\ client_encoding
Specify client encoding for database connection.

.TP
.BR -m,\ --module\ db_module
Specify which module to use for connection to DB server. Currently supported values are 'pq' for PostgreSQL, 'mysql' for MySQL, 'sqlite' for SQLite3 and 'file'. File module is not supported in Windows.

.TP
.BR -l,\ --libdir\ path_to_db_module
Specify path to DB modules. If you specify pq module, smsd searches in libdir for file libsmsd_pq.so (for mysql - libsmsd_mysql.so).
Do not interchange these plugins with default SQL libraries from database distribution!

.TP
.BR -f,\ --logfile\ file
If used, smsd will monitor sms traffic to file "file". File is
reopened and closed for every message so you can simply rename or
delete it if you want to rotate logs. If you use '-' as filename smsd
will use standard output.

.TP
.BR -t,\ --phone\ phone_number
You can run more instances of smsd. One instance per phone. In this
setup you must configure phone section in the config file. For example:

.PD 0
.RS
[global]
.P
port = /dev/ttyS0
.P
model = 6110
.P
initlength = default
.P
connection = serial
.P
bindir = /usr/local/sbin/
.P
use_locking = no
.P
serial_baudrate = 38400
.P
smsc_timeout = 10
.P

[phone_1]
.P
port = /dev/ttyS0
.P
model = 6110
.P
initlength = default
.P
connection = serial
.P
bindir = /usr/local/sbin/
.P
use_locking = no
.P
serial_baudrate = 38400
.P

[phone_2]
.P
port = /dev/ttyS1
.P
model = AT
.P
initlength = default
.P
connection = serial
.P
bindir = /usr/local/sbin/
.P
use_locking = no
.P
serial_baudrate = 38400
.P
smsc_timeout = 10
.PD

In this setup if you run smsd without \-t option it will use config
from global section. If you run smsd with \-t 1 it will use config
from phone_1 and likewise with \-t 2 uses phone_2. In case of use \-t
option smsd fills phone number for incoming sms's and in outbox it
looks only for sms's with the specified phone number in column
\&'phone'. This behaviour gives you possibility to run one instance
of smsd for every phone you have at the same time.
.RE

.TP
.BR -i,\ --interval\ interval
Polling interval for incoming sms's in seconds. Default is 1 second.

.TP
.BR -S,\ --maxsms\ number
When phone driver doesn't support sms status it is possible to specify
number of sms's supported by phone. Default is 10. If driver support
sms status this option is ignored.

.TP
.BR -b,\ --inbox\ memory_type
Where to look for new messages (defaults to IN for most Nokia phones and to
SM for all other phones).
For debugging or for unusual setups you can use another memory type
supported by your phone (eg. IN for the Inbox, OU for the Outbox, AR for the
Archive, TE for the Templates and F1,...,F20 for your own folders).
Use gnokii \-\-showsmsfolderstatus for a list of such memory types.

.TP
.BR -0,\ --firstpos0
SMSD starts messages reading from position 1. If your phone stores sms's 
from position 0, use this option.

.TP
.BR -v,\ --version
Shows smsd version.

.TP
.BR -h,\ --help
Shows short help.

.SH "DIAGNOSTICS"
Various error messages are printed to STDERR.

Debug output can be enabled setting xdebug=on (for messages specific to \fBSMSD\fR)
and debug=on (for messages specific to \fBlibgnokii\fR) in the configuration file.

.SH "BUGS"
.PP 
None reported.

.SH "AUTHOR"
\fBSMSD\fR was written by Jan Derfinak and contributed by other people.
See also Docs/CREDITS from the Gnokii sources.

.SH "COPYING"
This program is distributed under the GNU General Public License.

.SH "SEE ALSO"
gnokii(1)


ChangeLog
0.6.32 (not released yet)
=========================
 * translation updates
    o updated Italian translation                   (Daniele Forsi)
    o updated French translation                     (Raphaël Droz)
 * libgnokii updates
    o replace gethostbyname (deprecated) with getaddrinfo in tcp
      driver                                            (Paweł Kot)
    o add optional authentication during the initialization phase
                                                        (Paweł Kot)
    o move device_script() to generic device; it will cause ability
      to execute connect_script and disconnect_script with all
      connection types                                  (Paweł Kot)
    o implement GN_OP_Ping operation                    (Paweł Kot)
    o move callbacks from gn_data to separate structure put into
      gn_statemachine                                   (Paweł Kot)
    o add id for Nokia 3720c to misc.c only; use model=series40
                                                     (Anders Jenbo)
    o add another id for Nokia 1110 and ids for Nokia 1112 and 1208
      to misc.c only; use model=series40; many commands don't work,
      see the Series 30 section in Docs/README-6510
                                                   (Gustavo Conrad)
    o remove inline from the local_atoi() function definition and
      add static instead                                (Paweł Kot)
    o don't overwrite user-provided TP-Validity-Period
      in sms_prepare()                              (Daniele Forsi)
    o fix false detection of compressed SMS         (Daniele Forsi)
    o add id for Nokia 100; use model=series40; many commands don't
      work, see the Series 30 section in Docs/README-6510
                                                   (Bruno Estieule)
    o add GN_OP_Passthrough                         (Daniele Forsi)
 * at driver updates
    o in the default case autodetect if PDU SMS starts with SMSC
                                                    (Daniele Forsi)
    o fix parsing responses contining non-ascii characters
                                                        (Paweł Kot)
    o fix sending sms with Nokia Cellular Data Adapter; thanks to
      Tristan Drinkwater for patient testing            (Paweł Kot)
    o add optional authentication during the initialization phase
                                                        (Paweł Kot)
    o implement GN_OP_Ping operation                    (Paweł Kot)
    o don't overwrite user-provided TP-Validity-Period; this means
      that the default now is 3 days, like in other drivers,
      instead of 4 days forced by the driver        (Daniele Forsi)
 * fake driver updates
    o always autodetect if PDU SMS starts with SMSC (Daniele Forsi)
    o make it usable with gnome-phone-manager       (Daniele Forsi)
    o set TP-Validity-Period and update the testsuite to test it
                                                    (Daniele Forsi)
    o rewrite fake phonebook handling to work also on non C99
      compilers                                     (Daniele Forsi)
    o implement GN_OP_DeletePhonebook               (Daniele Forsi)
 * documentation updates
    o samples for connect and disconnect scripts for Portech
      devices                                   (Guillermo Benitez)
    o update man page with allowed values for --validity option of
      --sendsms                                     (Daniele Forsi)
 * nk6510 driver updates
    o store also the seconds in date fields when reading phonebook
      entries from phone                            (Daniele Forsi)
    o add support for phonebook memories SD (Service Dialling
      Numbers), MR (Message Recipients), CL (Call Log) and
      CB (Currently selected memory)                (Daniele Forsi)
    o parse call duration subblock found in Dialled Calls (DC),
      Received Calls (RC) and Call Log (CL) phonebook entries;
      output format is subject to change            (Daniele Forsi)
    o change the value used for phonebook memory ON (reading and
      deleting work, but writing fails)             (Daniele Forsi)
    o change the value used for VOICE phonebook memory, containing
      the phone number(s) to be used to listen to voicemail, which
      was used for MT which now is unsupported      (Daniele Forsi)
    o use libusb to detach cdc_phonet or any other linux kernel
      module; modprobe blacklisting isn't needed anymore
                                                    (Daniele Forsi)
    o write phone number read from TEL;TYPE=PREF,VOICE of a vCard
      when there is no other number in subentries   (Daniele Forsi)
    o implement GN_OP_Passthrough operation         (Daniele Forsi)
    o implement GN_OP_PollSMS operation (copied form nk6100.c)
                                                    (Daniele Forsi)
    o make GN_OP_PressPhoneKey and GN_OP_ReleasePhoneKey work on
      Series 40 3rd                                 (Daniele Forsi)
 * xgnokii updates
    o fix bug that caused to forget subentries after the first one
      when editing the phonebook                    (Daniele Forsi)
    o change relevant inline defintion to static inline to follow
      C99 standard                                      (Paweł Kot)
 * nk7110 driver updates
    o return more specific error codes for GN_OP_ReadPhonebook and
      for GN_OP_GetSMSCenter failures               (Daniele Forsi)
    o add support for phonebook memory SD (Service Dialling
      Numbers) and CB (Currently selected memory)   (Daniele Forsi)
    o change the value used for VOICE phonebook memory, containing
      the phone number(s) to be used to listen to voicemail, which
      was used for MT which now is unsupported      (Daniele Forsi)
    o write phone number read from TEL;TYPE=PREF,VOICE of a vCard
      when there is no other number in subentries   (Daniele Forsi)
 * gnokii updates
    o fix --getfile --getfilebyid --getallfiles and --getmms on

smsd/smsd.c
/*

  S M S D

  A Linux/Unix tool for the mobile phones.

  This file is part of gnokii.

  Copyright (C) 1999 Pavel Janik ml., Hugh Blemings
  Copyright (C) 1999-2011 Jan Derfinak

*/

#include "config.h"

#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <stdlib.h>
#include <pthread.h>
#include <getopt.h>
#include <time.h>

#ifndef WIN32
# include <unistd.h>  /* for usleep */
# include <signal.h>
#else
# include <windows.h>
# undef IN
# undef OUT
#endif

#include <glib.h>
#include <gmodule.h>

#include "misc.h"

#include "gnokii.h"
#include "smsd.h"
#include "lowlevel.h"
#include "db.h"

#ifdef ENABLE_NLS
#  include <locale.h>
#endif


/* Hold main configuration data for smsd */
SmsdConfig smsdConfig;

/* Global variables */
bool GTerminateThread;

/* Local variables */
static DBConfig connection;
void (*DB_Bye) (void) = NULL;
gint (*DB_ConnectInbox) (const DBConfig) = NULL;
gint (*DB_ConnectOutbox) (const DBConfig) = NULL;
gint (*DB_InsertSMS) (const gn_sms * const, const gchar * const) = NULL;
gint (*DB_Look) (const gchar * const) = NULL;

static pthread_t db_monitor_th;
pthread_mutex_t db_monitorMutex;
static volatile bool db_monitor;


gint LoadDB (void)
{
  GModule *handle;
  gchar *buf;
  gchar *full_name;

  full_name = g_strdup_printf ("smsd_%s", smsdConfig.dbMod);
  buf = g_module_build_path (smsdConfig.libDir, full_name);
  g_free (full_name);
  
  gn_log_xdebug ("Trying to load module %s\n", buf);
    
  handle = g_module_open (buf, G_MODULE_BIND_LAZY);
  g_free (buf);
  if (!handle)
  {
    g_print ("g_module_open error: %s!\n", g_module_error ());
    return (1);
  }
    
  if (g_module_symbol (handle, "DB_Bye", (gpointer *)&DB_Bye) == FALSE)
  {
    g_print ("Error getting symbol 'DB_Bye': %s\n", g_module_error ());
    return (2);
  }

  if (g_module_symbol (handle, "DB_ConnectInbox", (gpointer *)&DB_ConnectInbox) == FALSE)
  {
    g_print ("Error getting symbol 'DB_ConnectInbox':  %s\n", g_module_error ());
    return (2);
  }

  if (g_module_symbol (handle, "DB_ConnectOutbox", (gpointer *)&DB_ConnectOutbox) == FALSE)
  {
    g_print ("Error getting symbol 'DB_ConnectOutbox': %s\n", g_module_error ());
    return (2);
  }

  if (g_module_symbol (handle, "DB_InsertSMS", (gpointer *)&DB_InsertSMS) == FALSE)
  {
    g_print ("Error getting symbol 'DB_InsertSMS': %s\n", g_module_error ());
    return (2);
  }

  if (g_module_symbol (handle, "DB_Look", (gpointer *)&DB_Look) == FALSE)
  {
    g_print ("Error getting symbol 'DB_Look': %s\n", g_module_error ());
    return (2);
  }

  return (0);
}


static void Short_Version (void)
{
  g_print (_("smsd - version %s from gnokii %s\n"), SVERSION, VERSION);
}

static void Version (void)
{
  g_print (_("smsd - version %s from gnokii %s\nCopyright  Jan Derfinak 1999-2011\n"), SVERSION, VERSION);
}


static void Usage (gchar *p)
{
  Version ();
  g_print (_("\nUsage:  %s [options]\n"
             "            -u, --user db_username OR action if -m file\n"
             "            -p, --password db_password\n"
             "            -d, --db db_name\n"
             "            -c, --host db_hostname OR spool directory if -m file\n"
             "            -s, --schema db_schema\n"
             "            -e, --encoding client_encoding\n"
             "            -m, --module db_module (pq, mysql, sqlite, file)\n"
             "            -l, --libdir path_to_db_module\n"
             "            -C, --config path_to_config_file\n"
             "            -f, --logfile file\n"
             "            -t, --phone phone_number\n"
             "            -i, --interval polling_interval_for_incoming_sms's_in_seconds\n"
             "            -S, --maxsms number_of_sms's (only in dumb mode)\n"
             "            -b, --inbox memoryType\n"
             "            -0, --firstpos0\n"
             "            -v, --version\n"
             "            -h, --help\n"), p);
}


static void LogFile (gchar *str, ...)
{
  FILE *f;
  va_list ap;
  time_t cas;
  gchar buf[50];
  
  if (smsdConfig.logFile == NULL)
    return;

  if (strcmp (smsdConfig.logFile, "-") == 0)
    f = stdout;
  else if ((f = fopen (smsdConfig.logFile, "a")) == NULL)
  {
    g_print (_("WARNING: Cannot open file %s for appending.\n"), smsdConfig.logFile);
    return;
  }

  cas = time (NULL);
  strftime (buf, 50, "%e %b %Y %T", localtime (&cas));
  fprintf (f, "%s: ", buf);

  va_start (ap, str);
  vfprintf (f, str, ap);
  va_end (ap);
  if (f == stdout)
    fflush (f);
  else
    fclose (f);
  return;
}


static void ReadConfig (gint argc, gchar *argv[])
{
  connection.user = g_strdup ("");
  connection.password = g_strdup ("");
  connection.db = g_strdup ("sms");
  connection.host = g_strdup ("");
  connection.schema = g_strdup ("public");
  connection.clientEncoding = g_strdup ("");
  smsdConfig.dbMod = g_strdup ("file");
  smsdConfig.libDir = g_strdup (MODULES_DIR);
  smsdConfig.configFile = NULL;
  smsdConfig.logFile = NULL;
  smsdConfig.phone = g_strdup ("");
  smsdConfig.refreshInt = 1;     // Phone querying interval in seconds
  smsdConfig.maxSMS = 10;        // Smsd uses it if GetSMSStatus isn't implemented
  smsdConfig.firstSMS = 1;
//  smsdConfig.smsSets = 0;
  smsdConfig.memoryType = GN_MT_XX;

  while (1)
  {
    gint optionIndex = 0;
    gint c;
    static struct option longOptions[] =
    {
      {"user", 1, 0, 'u'},
      {"password", 1, 0, 'p'},
      {"db", 1, 0, 'd'},
      {"host", 1, 0, 'c'},
      {"schema", 1, 0, 's'},
      {"encoding", 1, 0, 'e'},
      {"module", 1, 0, 'm'},
      {"libdir", 1, 0, 'l'},
      {"config", 1, 0, 'C'},
      {"logfile", 1, 0, 'f'},
      {"phone", 1, 0, 't'},
      {"version", 0, 0, 'v'},
      {"interval", 1, 0, 'i'},
      {"maxsms", 1, 0, 'S'},
      {"inbox", 1, 0, 'b'},
      {"firstpos0", 0, 0, '0'},
      {"help", 0, 0, 'h'},
      {0, 0, 0, 0}
    };
    
    c = getopt_long (argc, argv, "u:p:d:c:s:e:m:l:C:f:t:vi:S:b:0h", longOptions, &optionIndex);
    if (c == EOF)
      break;
    switch (c)
    {
      case 'u':
        g_free (connection.user);
        connection.user = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;
      
      case 'p':
        g_free (connection.password);
        connection.password = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;
        
      case 'd':
        g_free (connection.db);
        connection.db = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;
        
      case 'c':
        g_free (connection.host);
        connection.host = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;
      
      case 's':
        g_free (connection.schema);
        connection.schema = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;

      case 'e':
        g_free (connection.clientEncoding);
        connection.clientEncoding = g_strdup (optarg);
        memset (optarg, 'x', strlen (optarg));
        break;
        
      case 'm':
        g_free (smsdConfig.dbMod);
        smsdConfig.dbMod = g_strdup (optarg);
        break;
        
      case 'l':
        g_free (smsdConfig.libDir);
        smsdConfig.libDir = g_strdup (optarg);
        break;

      case 'C':
        if (smsdConfig.configFile)
          g_free (smsdConfig.configFile);
        smsdConfig.configFile = g_strdup (optarg);
        break;

      case 'f':
        if (smsdConfig.logFile)
          g_free (smsdConfig.logFile);
        smsdConfig.logFile = g_strdup (optarg);
        break;

      case 't':
        if (smsdConfig.phone)
          g_free (smsdConfig.phone);
        smsdConfig.phone = g_strdup (optarg);
        break;

      case 'i':
        smsdConfig.refreshInt = atoi (optarg);
        break;

      case 'S':
        smsdConfig.maxSMS = atoi (optarg);
        break;

      case 'b':
        smsdConfig.memoryType = gn_str2memory_type (optarg);
        break;

      case '0':
        smsdConfig.firstSMS = 0;
        break;
        
      case 'v':
        Version ();
        exit (0);
        break;

      case 'h':
      case '?':
        Usage (argv[0]);
        exit (1);
        break;

      default:
        g_print ("getopt returned 0%o\n", c);
    }
  }
  
  if ((argc - optind) != 0)
  {
    g_print (_("Wrong number of arguments\n"));
    Usage (argv[0]);
    exit (1);
  }

  if (LoadDB ())
  {
    g_print (_("Cannot load database module %s in directory %s!\n"),
              smsdConfig.dbMod, smsdConfig.libDir);
    exit (-2);
  }
}


static void *SendSMS (void *a)
{
  if ((*DB_ConnectOutbox) (connection))
  {
    pthread_exit (0);
    return (NULL);
  }

  while (1)
  {
    pthread_mutex_lock (&db_monitorMutex);
    if (!db_monitor)
    {
      pthread_mutex_unlock (&db_monitorMutex);
      pthread_exit (0);
      return (NULL);
    }
    pthread_mutex_unlock (&db_monitorMutex);

    if ((*DB_Look) (smsdConfig.phone) == SMSD_OUTBOXEMPTY)
      sleep (5);
  }
}


static void MainExit (gint sig)
{
  PhoneEvent *e = (PhoneEvent *) g_malloc (sizeof (PhoneEvent));

  e->event = Event_Exit;
  e->data = NULL;
  InsertEvent (e);
  
  pthread_mutex_lock (&db_monitorMutex);
  db_monitor = FALSE;
  pthread_mutex_unlock (&db_monitorMutex);
  
  pthread_join (monitor_th, NULL);
  pthread_join (db_monitor_th, NULL);
  (*DB_Bye) ();
  exit (sig);
}


GNOKII_API gint WriteSMS (gn_sms *sms)
{
  gn_error error;
  PhoneEvent *e = (PhoneEvent *) g_malloc (sizeof (PhoneEvent));
  D_SMSMessage *m = (D_SMSMessage *) g_malloc (sizeof (D_SMSMessage));

  m->sms = sms;
  e->event = Event_SendSMSMessage;
  e->data = m;
  InsertEvent (e);
  pthread_mutex_lock (&sendSMSMutex);
  pthread_cond_wait (&sendSMSCond, &sendSMSMutex);
  pthread_mutex_unlock (&sendSMSMutex);

  gn_log_xdebug ("Address: %s\nText: %s\n", sms->remote.number, sms->user_data[0].u.text);

  error = m->status;
  g_free (m);

  if (error)
  {
    gn_log_xdebug ("Sending to %s unsuccessful. Error: %s\n", sms->remote.number, gn_error_print(error));
    if (smsdConfig.logFile)
      LogFile (_("Sending to %s unsuccessful. Error: %s\n"), sms->remote.number, gn_error_print(error));
  }
   else
  {
    gn_log_xdebug ("Sending to %s successful.\n", sms->remote.number);
    if (smsdConfig.logFile)
      LogFile (_("Sending to %s successful.\n"), sms->remote.number);
  }

  return (error);
}


static void ReadSMS (gpointer d, gpointer userData)
{
  gn_sms *data = (gn_sms *)d;
  PhoneEvent *e;
  gint error;

  if (data->type == GN_SMS_MT_Deliver || data->type == GN_SMS_MT_StatusReport)
  {
    gn_log_xdebug ("%d. %s   ", data->number, data->remote.number);
    gn_log_xdebug ("%02d-%02d-%02d %02d:%02d:%02d+%02d %s\n", data->smsc_time.year,
                    data->smsc_time.month, data->smsc_time.day, data->smsc_time.hour,
                    data->smsc_time.minute, data->smsc_time.second, data->smsc_time.timezone,
                    data->user_data[0].u.text);
    error = (*DB_InsertSMS) (data, smsdConfig.phone);

    switch (error) 
    {
      case SMSD_OK:
        if (smsdConfig.logFile)
          LogFile (_("Inserting sms from %s successful.\n"), data->remote.number);
        gn_log_xdebug ("Inserting sms from %s successful.\n", data->remote.number);
        e = (PhoneEvent *) g_malloc (sizeof(PhoneEvent));
        e->event = Event_DeleteSMSMessage;
        e->data = data;
        InsertEvent (e);
        break;
        
      case SMSD_DUPLICATE:
        if (smsdConfig.logFile)
          LogFile (_("Duplicated sms from %s.\n"), data->remote.number);
        gn_log_xdebug ("Duplicated sms from %s.\n", data->remote.number);
        e = (PhoneEvent *) g_malloc(sizeof(PhoneEvent));
        e->event = Event_DeleteSMSMessage;
        e->data = data;
        InsertEvent (e);
        break;

      case SMSD_WAITING:
        if (smsdConfig.logFile)
          LogFile (_("Multipart sms from %s. Waiting for the next parts.\n"), data->remote.number);
        gn_log_xdebug ("Multipart sms from %s. Waiting for the next parts.\n", data->remote.number);
        e = (PhoneEvent *) g_malloc (sizeof(PhoneEvent));
        e->event = Event_DeleteSMSMessage;
        e->data = data;
        InsertEvent (e);
        break;
        
      default:
        if (smsdConfig.logFile)
          LogFile (_("Inserting sms from %s unsuccessful.\nDate: %02d-%02d-%02d %02d:%02d:%02d+%02d\nText: %s\n\nExiting.\n"),
                   data->remote.number, data->smsc_time.year,
                   data->smsc_time.month, data->smsc_time.day,
                   data->smsc_time.hour, data->smsc_time.minute,
                   data->smsc_time.second, data->smsc_time.timezone,
                   data->user_data[0].u.text);
        gn_log_xdebug ("Inserting sms from %s unsuccessful.\nDate: %02d-%02d-%02d %02d:%02d:%02d+%02d\nText: %s\n\nExiting.\n",
                        data->remote.number, data->smsc_time.year,
                        data->smsc_time.month, data->smsc_time.day,
                        data->smsc_time.hour, data->smsc_time.minute,
                        data->smsc_time.second, data->smsc_time.timezone,
                        data->user_data[0].u.text);
        break;
    }
  }
}


static void GetSMS (void)
{
  while (1)
  {
    pthread_mutex_lock (&smsMutex);
    // Waiting for signal
    pthread_cond_wait (&smsCond, &smsMutex);

    // Signal arrived.
    pthread_mutex_unlock (&smsMutex);

    // Check if Mutex is free
    pthread_mutex_lock (&smsMutex);
    phoneMonitor.sms.messages = g_slist_reverse(phoneMonitor.sms.messages);
    g_slist_foreach (phoneMonitor.sms.messages, ReadSMS, (gpointer) NULL);
    pthread_mutex_unlock (&smsMutex);
  }
}


static void Run (void)
{
    /* Windows doesn't support signaling */
#ifndef WIN32
  struct sigaction act;
  
  act.sa_flags = 0;
  act.sa_handler = MainExit;
  sigemptyset (&(act.sa_mask));
  
  sigaction (SIGQUIT, &act, NULL);
  sigaction (SIGTERM, &act, NULL);
  sigaction (SIGINT, &act, NULL);

#if __unices__
  act.sa_handler = SIG_IGN;
  sigemptyset (&(act.sa_mask));
  sigaction (SIGALRM, &act, NULL);
#endif
#endif

  InitPhoneMonitor ();
  if ((*DB_ConnectInbox) (connection))
    exit (2);
  pthread_create (&monitor_th, NULL, Connect, smsdConfig.phone);
  db_monitor = TRUE;
  pthread_mutex_init (&db_monitorMutex, NULL);
  pthread_create (&db_monitor_th, NULL, SendSMS, NULL);
  GetSMS ();
}


int main (int argc, char *argv[])
{
#ifdef ENABLE_NLS
  setlocale(LC_ALL, "");
  bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);
  textdomain(GETTEXT_PACKAGE);
#endif

  gn_elog_handler = NULL;
  ReadConfig (argc, argv);
  /* Introduce yourself */
  Short_Version ();
  GTerminateThread = false;
  Run ();

  return(0);
}
